File: frontend/flutter_app/lib/models/student.dart
class SubjectResult {
  SubjectResult({required this.id, required this.name, required this.grade});

  factory SubjectResult.fromJson(Map<String, dynamic> json) => SubjectResult(
        id: json['_id'] as String?,
        name: json['name'] as String? ?? '',
        grade: json['grade'] as String? ?? '',
      );

  String? id;
  String name;
  String grade;

  Map<String, dynamic> toJson() => {'name': name, 'grade': grade};
}

class Student {
  Student({
    required this.id,
    required this.fullName,
    required this.email,
    this.school = '',
    this.district = '',
    this.alYear,
    this.stream,
    this.dateOfBirth,
    this.alResults = const [],
    this.profileCompletion = 0,
    this.personalityCategory,
  });

  factory Student.fromJson(Map<String, dynamic> json) {
    final personality = json['personalityResult'] as Map<String, dynamic>?;
    return Student(
      id: json['_id'] as String,
      fullName: json['fullName'] as String? ?? '',
      email: json['email'] as String? ?? '',
      school: json['school'] as String? ?? '',
      district: json['district'] as String? ?? '',
      alYear: (json['alYear'] as num?)?.toInt(),
      stream: json['stream'] as String?,
      dateOfBirth: json['dateOfBirth'] as String?,
      alResults: ((json['alResults'] as List?) ?? const [])
          .map((item) => SubjectResult.fromJson(item as Map<String, dynamic>))
          .toList(),
      profileCompletion: (json['profileCompletion'] as num?)?.toInt() ?? 0,
      personalityCategory: personality?['category'] as String?,
    );
  }

  final String id;
  String fullName;
  final String email;
  String school;
  String district;
  int? alYear;
  String? stream;
  String? dateOfBirth;
  List<SubjectResult> alResults;
  int profileCompletion;
  String? personalityCategory;

  Map<String, dynamic> toJson() => {
        'fullName': fullName,
        'school': school,
        'district': district,
        'alYear': alYear,
        'stream': stream,
        'dateOfBirth': dateOfBirth,
        'alResults': alResults.map((result) => result.toJson()).toList(),
      };
}

File: backend/models/Student.js
const mongoose = require('mongoose');

const subjectResultSchema = new mongoose.Schema(
	{
		name: { type: String, required: true, trim: true, maxlength: 80 },
		grade: { type: String, required: true, trim: true, maxlength: 10 },
	},
	{ _id: true },
);

const studentSchema = new mongoose.Schema(
	{
		userId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true, unique: true, index: true },
		fullName: { type: String, required: true, trim: true, maxlength: 100 },
		email: { type: String, required: true, lowercase: true, trim: true },
		dateOfBirth: { type: Date },
		school: { type: String, trim: true, maxlength: 150 },
		district: { type: String, trim: true, maxlength: 80 },
		alYear: { type: Number, min: 1900, max: 2200 },
		stream: { type: String, enum: ['Science', 'Mathematics', 'Commerce', 'Arts', 'Technology'] },
		profileImage: { type: String, trim: true },
		interests: [{ type: String, trim: true }],
		alResults: [subjectResultSchema],
		personalityResult: {
			category: { type: String, trim: true },
			strengths: [{ type: String, trim: true }],
			completedAt: { type: Date },
		},
		favoriteCareers: [{ type: mongoose.Schema.Types.ObjectId }],
		favoriteCourses: [{ type: mongoose.Schema.Types.ObjectId }],
	},
	{ timestamps: true, toJSON: { virtuals: true } },
);

studentSchema.virtual('profileCompletion').get(function getProfileCompletion() {
	const checks = [
		Boolean(this.fullName), Boolean(this.email), Boolean(this.school), Boolean(this.district),
		Boolean(this.alYear), Boolean(this.stream), this.alResults.length > 0,
		Boolean(this.personalityResult?.category),
	];
	return Math.round((checks.filter(Boolean).length / checks.length) * 100);
});

module.exports = mongoose.model('Student', studentSchema);
